### PR TITLE
Fix css injection on Chrome [improved but not perfect]

### LINKED ISSUE #3418: Custom CSS for Link Hints inside iFrames seems to not be working correctly again
I noticed this has been referenced before.
#2594
And fixed by:
#2683

However, it is still occurring for me. First started noticing it today [10.28.19].

Chrome Version: 78.0.3904.70 (Official Build) (64-bit) (cohort: 78_70_Win)
Vimium Version: 1.64.6

### LINKED ISSUE #3452: Custom CSS not applied to new pages
I have some custom CSS set (which is a much appreciated feature) but there are times when the default styles are still applied until I reload the page.

Steps to reproduce the behavior (90% consistent):
1. Press 't' to open a new tab
2. Type in a search and hit enter
3. Press 'f' for link hints

Since it only seems to happen on a new page and not every single time, maybe there's something happening while things are loading or initializing? 

With custom CSS:
Screen Shot 2020-01-01 at 8 33 16 PM

Custom CSS not being applied:
Screen Shot 2020-01-01 at 8 32 53 PM

Chrome: 79.0.3945.88
Vimium: 1.64.6
macOS: 10.15.2

### LINKED ISSUE #3956: Custom CSS Set In Vimium Options Not Rendering On Webpages
**Describe the bug**
Include a clear bug description.

I set the CSS in the "CSS for Vimium UI" box. However, some pages seem to recognize the change of color from yellow to blue, while others do not.

**To Reproduce**
1. Go to URL '...'

Include a screenshot if applicable.

image

image

image

**Browser and Vimium version**

If you're using Chrome, include the Chrome and OSS version found at chrome://version. Also include the Vimium version found at chrome://extensions.

If you're using Firefox, report the Firefox and OS version found at about:support. Also include the Vimium version found at about:addons.

Google Chrome - Version 95.0.4638.69 (Official Build) (x86_64)
I'm on Vimium 1.67

### PR DESCRIPTION
**Pull request by gdh1995**
This ensures `userDefinedLinkHintCss` can override those in `vimium.css`, so it can replace #3421, fix #3418, fix #3452 and fix #3956.

As said in https://github.com/philc/vimium/issues/3418#issuecomment-549160351 , during my tests the document about `chrome.tabs.insertCSS` is not correct - the CSS may be used earlier than those defined in `manifest.json`.

Updated (2021/11/16): as said in #3956, CSS inserted at `document_start` may be used earlier even on a top iframe on Chrome 95+Win10.

### CODE REVIEW

**Comment by fiskhest** [comment]
Aching to see this fix implemented in master. Great stuff @gdh1995 !

**Comment by hisener** [comment]
Tested on Chrome 94.0.4606.54 by installing from the source. It works like a charm! Thank you. :pray:

**Comment by BachoSeven** [comment]
Also tested on Chromium 94.0.4606.71 and it resolves all race conditions. Thanks @gdh1995 for the fix!

**Comment by philiparvidsson** [comment]
What's the status here?

**Comment by ceiphr** [comment]
Any updates on when this might be merged? I'm still experiencing #3956.

**Comment by philiparvidsson** [comment]
How is this not merged yet? It's a super annoying bug. I've inserted custom CSS to have blank link hints with white text. But now I get the yellow default hints with white text, making them barely readable.

**Comment by austintraver** [comment]
I hope a maintainer with merge access sees this PR. This code fixes https://github.com/philc/vimium/issues/3956

File: background_scripts/main.js [hunk 1]
```diff
@@ -400,7 +400,15 @@ chrome.webNavigation.onCommitted.addListener(function({tabId, frameId}) {
     code: Settings.get("userDefinedLinkHintCss"),
     runAt: "document_start"
   };
-  return chrome.tabs.insertCSS(tabId, cssConf, () => chrome.runtime.lastError);
+  const callback = () => chrome.runtime.lastError;
+  chrome.tabs.insertCSS(tabId, cssConf, callback);
+  if (Utils.isFirefox()) {
```
Comment (alex35mil) [inline background_scripts/main.js:405]: FWIW I see the same issue in Firefox.

**Comment by philc** [comment]
Sorry for the delay here. I'd like to merge this, but I haven't been able to reproduce the underlying issue. Is there a reliable test case?

**Comment by philc** [comment]
OK, I was finally able to reproduce this. Thank you @gdh1995. My notes:

* I couldn't reproduce this in MacOS, only windows.
* It reliably occurred once in about ten tries. Load "C:/" as the URL in windows. Hit reload, and then press `f` to activate link hints and see if the custom styles appear.
* I never saw this occur in Firefox in Windows or MacOS using the same procedure.

I also couldn't reproduce this in Vimium 2.0 on either Windows or MacOS. This may be because we're now using the chrome.scripting.insertCSS instead of chrome.tabs.insertCSS (chrome.tabs.insertCSS was removed in Manifest v3), and chrome.scripting.insertCSS somehow behaves differently such that this issue doesn't occur.

I'm going to close this as I believe it should finally be resolved once Vimium 2.0 hits the store.

If anyone is still able to reproduce this on Vimium 2.0, please post a comment indicating the procedure to do so, and your browser and OS.

**Pull request closed without merging**